text/0000-ember-data-identifiers.md

```diff
@@ -0,0 +1,396 @@
+  to calling this method. We leave it to follow up RFCs to provide infrastructure for safely and correctly eliminating records from caches.
+
+```ts
+Store {
+  forgetRecordIdentifier(identifier: Identifier): void
+}
+```
+
+### Access from `record` instances
+
+```typescript
+export function recordIdentifierFor(record: object): Identifier {}
+```
+
+```ts
+import { recordIdentifierFor } from '@ember-data/store';
+
+// ...
+
+// when you have a record instance
+let identifier = recordIdentifierFor(record);
+
+// from inside a record class after instantiation
+class MyRecord {
+  getIdentifier() {
+    return recordIdentifierFor(this);
+  }
+}
+```
+
+Whether an how to accessing an identifier during instantiation of a record will be left for discussion as part
+  of a different RFC for custom record classes.
+
+### Polymorphism & "The Username Problem"
+
+A common edge case that `Identifiers` enables end users to solve is when multiple pieces of identifying information should reference
+ the same data.
+ 
+For instance, when using single-table polymorphism (in which `ferrari` and `bmw` extend `car` and share a common `id` space)
+ then `ferrari:1` and `bmw:2` are the same vehicles as `car:1` and `car:2`.
+
+A similar problem presents for the scenario in which we know that we wish to reference a `user` with a given `username`, but do not yet
+ have access to the `id` for that user. In this case, `user:@jackson5` and `user:abc123` are the same user.
+
+Today in these situations many users will encounter bugs resulting from there being two records present in the cache instead of one. This
+ problem can be solved with a custom identifier generation method that is aware of an application's polymorphic associations or additional
+ indexing requirements.
+
+ For example, to solve the username problem we might do the following:
+
+ ```typescript
+import { setIdentifierGenerationMethod } from '@ember-data/store';
+
+let count = 0;
+const typeid_cache = {};
+const username_cache = {};
+
+setIdentifierGenerationMethod((resource: Resource) => {
+  let { type, id, lid } = resource;
+  let username = resource.type === 'user' && resource.attributes && resource.attributes.username;
+  let cacheKey, altCacheKey;
+
+  if (lid) {
+    // probably ensure username and id cache are populated first IRL
+    return lid;
+  }
+
+  // handle the case where we do know the ID and have set the `lid` previously
+  if (id) {
+    cacheKey = `${type}:${id}`;
+
+    if (lid = typeid_cache[cacheKey]) {
+      return lid;
+    }
+  }
+
+  // handle the cases where we have a username but we didn't know the ID yet
+  if (username) {
+    lid = username_cache[username];
+
+    if (!lid) {
+      lid =  `my-key-${count++}`;
+      username_cache[username] = lid;
+    }
+
+    if (id) {
+      typeid_cache[cacheKey] = lid;
+    }
+    return lid;
+  }
+
+  // handle everything else
+  lid =  `my-key-${count++}`;
+  typeid_cache[cacheKey] = lid;
+
+  return lid;
+});
+ ```
```

Comment:
It would help to augment this example with store API to make it clear what the flow is.

eg if I'm doing

```js
setIdentifierGenerationMethod(exampleInRFC);

store.findRecord('user', 'jackson5').
  then(doSomething); // this makes a request

// later

store.findRecord('user', 'abc123').
  then(doSomethingWithUser); // this does not make a request
```

At what times does the store call the generation method and with what arguments?

Reply:
Users should do any identifier generation configuration within an instance initializer prior to making use of the store.

For `RecordIdentifiers` the method is called once a record (remote or local) becomes known about. In the `findRecord` case this will be greedy (prior to network request being sent), so as long as the first request for `jackson5` in the example had populated a secondary lookup table for `abc123` to point at the same identifier prior to the second request being fired no request would be made.

In cases such as `findRecord` or async relationships where `type+id` is known but full resources are not available, `updateRecordIdentifier` will be called once the resource is loaded to ensure that we populate these secondary lookup tables.

I'll add clarifying text around this.

Reply:
> In the findRecord case this will be greedy (prior to network request being sent), so as long as the first request for jackson5 in the example had populated a secondary lookup table for abc123 to point at the same identifier prior to the second request being fired no request would be made.

Indeed.  What causes that secondary cache population to happen?  Is your intent in the example above for that to happen out of band from a userland adapter?  If so, adding that detail will go a long way to making it clear how this example solves the issue of secondary caches.

Reply:
Added more details around secondary caches and expanded examples.